compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -13266,6 +13266,276 @@ OMR::Z::TreeEvaluator::arraysetEvaluator(TR::Node * node, TR::CodeGenerator * cg
    return NULL;
    }
 
+
+void 
+OMR::Z::TreeEvaluator::genLoopForwardArrayCopy(TR::Node *node, TR::CodeGenerator *cg, TR::Register *byteSrcReg, TR::Register *byteDstReg, TR::Register *loopIterReg, bool isConstantLength)
+   {
+   TR::LabelSymbol *processingLoop = generateLabelSymbol(cg);
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, processingLoop);
+   generateSS1Instruction(cg, TR::InstOpCode::MVC, node, 255, 
+      new (cg->trHeapMemory()) TR::MemoryReference(byteDstReg, 0, cg), 
+      new (cg->trHeapMemory()) TR::MemoryReference(byteSrcReg, 0, cg));
+   generateRXInstruction(cg, TR::InstOpCode::LA, node, byteSrcReg, new (cg->trHeapMemory()) TR::MemoryReference(byteSrcReg, 256, cg));
+   generateRXInstruction(cg, TR::InstOpCode::LA, node, byteDstReg, new (cg->trHeapMemory()) TR::MemoryReference(byteDstReg, 256, cg));
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRCT, node, loopIterReg, processingLoop);
+   // If it is known at compile time that this is forward array copy and also length is known at compile time then internal control flow starts here.
+   if (node->isForwardArrayCopy() && isConstantLength)
```

Comment:
Aren't we always a forward arraycopy in here, so why do we need to check that `node->isForwardArrayCopy()` is true?

Reply:
We are always will be generating forward array copy sequence here that is true. But ICF starts here only where we have constant byte length and node is known to be doing array copy in forward direction. If direction is not known and byteLength is constant, arraycopy ICF starts early when we check out overlapping region for array copy